csrc/dynamic_transform.cpp

```diff
@@ -0,0 +1,510 @@
+  std::vector<int64_t> inp_shape(inp_dom.size(), 0);
+  for (const auto i : c10::irange(inp_dom.size())) {
+    auto inp_id = inp_dom.at(i);
+    // This should have been validated when initially creating reshape
+    // op, but just in case
+    TORCH_INTERNAL_ASSERT(
+        !inp_id->maybePartial(),
+        "Invalid domain to reshape: ",
+        inp_id->toString());
+    auto extent_val = expr_eval_->evaluate(inp_id->extent());
+    TORCH_INTERNAL_ASSERT(
+        extent_val.has_value(),
+        "Cannot evaluate the extent of an input domain to reshape: ",
+        inp_id->toString());
+    TORCH_INTERNAL_ASSERT(
+        extent_val->isInt(),
+        "Invalid evaluated value of domain extent: ",
+        inp_id->toString());
+    TORCH_INTERNAL_ASSERT(
+        extent_val->as<int64_t>() > 0,
+        "Invalid input domain extent: ",
+        extent_val->as<int64_t>());
+    inp_shape.at(i) = extent_val->as<int64_t>();
+  }
+
+  std::cerr << "Input shape: " << inp_shape << std::endl;
+
+  const auto& out_dom = out_tv->getMaybeRFactorDomain();
+
+  // Determine output shape using expr evaluator. Note there may be
+  // one domain of extent -1
+  std::vector<int64_t> out_shape(out_dom.size(), 0);
+  bool extent_m1_found = false;
+  for (const auto i : c10::irange(out_dom.size())) {
+    auto out_id = out_dom.at(i);
+    auto extent_val = expr_eval_->evaluate(out_id->extent());
+    TORCH_INTERNAL_ASSERT(
+        extent_val.has_value(),
+        "Cannot evaluate the extent of an output domain to reshape: ",
+        out_id->toString());
+    TORCH_INTERNAL_ASSERT(
+        extent_val->isInt(),
+        "Invalid evaluated value of domain extent: ",
+        out_id->toString());
+    const auto extent_int = extent_val->as<int64_t>();
+    if (extent_int == -1) {
+      TORCH_INTERNAL_ASSERT(
+          !extent_m1_found,
+          "Multiple output domains of size -1 not allowed",
+          out_tv->toString());
+      extent_m1_found = true;
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          extent_int > 0, "Invalid output domain extent: ", extent_int);
+    }
+    out_shape.at(i) = extent_int;
+  }
+
+  std::cerr << "Output shape: " << out_shape << std::endl;
+
+  auto view_result = analyzeView(inp_tv, inp_shape, out_shape);
+
+  info_.reshape_transforms_.emplace_back(out_tv, view_result);
+}
+
+class TORCH_CUDA_CU_API DynamicTransformConcretizer : public OptOutMutator {
+ public:
+  DynamicTransformConcretizer(Fusion* fusion, const DynamicTransformInfo& info)
+      : info_(info) {
+    TORCH_INTERNAL_ASSERT(
+        fusion == info.fusion(),
+        "Invalid DynamicTransformInfo. The associated Fusion is different from the given Fusion");
+    concretize();
+  }
+
+ private:
+  void concretize();
+
+  void concretizeReshape();
+
+ private:
+  using OptOutMutator::mutate;
+
+  void mutate(TensorView* tv) final;
+  void mutate(TensorDomain* td) final;
+
+  bool propagateFromProducerToConsumer(TensorView* consumer);
```

Comment:
So this is because consumer iter types are based on input iter types. If there's a symbolic producer iter type that can propagate into consumers. I would assume at the point of a reduction/gather/etc the iter type would be specified correctly instead of inheriting symbolic type from poducer. So this is just to replace the instances where an iter domain is marked as symbolic after the reshapes have been concretized.